onnxruntime/core/providers/cuda/tensor/gather_elements_impl.cu

```diff
@@ -7,114 +7,138 @@
 namespace onnxruntime {
 namespace cuda {
 
-template <typename T, typename Tin>
+namespace {
+constexpr int threads_per_block = GridDim::maxThreadsPerBlock;
+constexpr int thread_worksize = 16;
```

Comment:
How was this value chosen ?

Reply:
Out of all questions, this is the one that needed to be asked. The short answer is that it was chosen experimentally. The original implicit value was 1. I noticed that some of our kernels, such as all BinaryElementWise ops have a value of 4.
I tried 4, 8 and 16 with diminishing returns. While the algo optimizations made things better, most of the gains were obtained from making a kernel process multiple index values. This sounds counter-intuitive, after all we expect the power of GPU parallel processing to easily handle millions of threads at the same time. Turns out it is not true. The number of microprocessors is limited and there are costs of lunching the kernel. I suspect most of the limitations come from the number of registers (32) available to a single kernel so they don't step on each other. While more trivial kernels run in parallel just fine, this one is less trivial and it does show its flaws when processing 10E6 elements at once. I have discussed my findings with Ke Deng and he confirmed that this is, in fact, what may be happening and making kernels process multiple items is a usual practice. (pytorch does it). Furthermore, he thinks that unrolling this extra loop does in fact work.

---
In reply to: [480482726](https://github.com/microsoft/onnxruntime/pull/4989#discussion_r480482726) [](ancestors = 480482726)